docs/agent-development.md

```diff
@@ -384,3 +387,54 @@ Here's a list of the config options across all agents, their types, default valu
+
+Agents should keep a record of the `Etag` value returned by the most successful config request, and then present it to future requests via the [If-None-Match](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match) header. If the config has not changed, the server will respond with 304 (Not Modified).
+
+The `Cache-Control` header should contain a `max-age` directive, specifying the amount of time (in seconds) the response should be considered "fresh". Agents should use this to decide how long to wait before requesting config again. The server will respond with a `Cache-Control` header even if the request fails.
+
+#### Dealing with errors
+
+Agents must deal with various error scenarios, including:
+
+ - 7.3 servers where the Kibana connection is not enabled (server responds with 403)
+ - 7.3 servers where the Kibana connection is enabled, but unavailable (server responds with 503)
+ - 7.3 servers where the Kibana connection is enabled, available, but there is no matching agent config (server responds with 404)
+ - pre-7.3 servers that don't support the config endpoint (server responds with 404)
+ - any other error (server responds with 5xx)
+
+If the server responds with any 5xx, agents should log at error level. If the server responds with 4xx, agents are not required to log the response. Either central config is not available, not enabled, or there is no agent config to update. In all of these cases, there is no expectation that the agent should do anything, so logging is not useful.
+
+In any case, a 7.3+ server _should_ respond with a Cache-Control header, as described in the section above, and agents should retry after the specified interval. For older servers, or for whatever reason a 7.3+ server does not respond with that header (or it is invalid), agents should retry after 5 minutes. We include this behaviour for older servers so that the agent will start polling after server upgrade without restarting the application.
+
+If the agent does not recognise a config attribute, or does not support dynamically updating it, then it should log a warning such as:
+
+```
+Central config failure. Unsupported config names: unknown_option, disable_metrics, capture_headers
+```
+
+Note that in the initial implementation of this feature, not all config attributes will be supported by the APM UI or APM Server. Agents may choose to support only the attributes supported by the UI/server, or they may choose to accept additional attributes. The latter will enable them to work without change once additional config attributes are supported by the UI/server.
+
+If the agent receives a known but invalid config attribute, it should log a warning such as:
+
+```
+Central config failure. Invalid value for transactionSampleRate: 1.2 (out of range [0,1.0])
```

Comment:
This and the other log warning also make sense for other config sources such as env variables.

Reply:
Agree, but we don't discuss that at all. Let's do that in a followup if you don't mind